drivers/flash/Kconfig.mcux

```diff
@@ -130,3 +130,32 @@ config FLASH_MCUX_XSPI
 	select FLASH_HAS_EXPLICIT_ERASE
 	help
 	  Enable the mcux xspi flash driver.
+
+if DT_HAS_NXP_S32K3X_C40_ENABLED
+# MCUX C40 internal flash driver (S32K3x)
+config SOC_FLASH_MCUX_C40
+	def_bool y
+	select FLASH_HAS_DRIVER_ENABLED
+	select FLASH_HAS_EXPLICIT_ERASE
+	select FLASH_HAS_PAGE_LAYOUT
+	# C40 uses controller commands, not CPU stores to flash
+	# so we are not relaxing MPU for ROM writes.
+	# Relocate driver/HAL when XIP so erase/program run from SRAM:
+	imply CODE_DATA_RELOCATION if XIP
+	imply CODE_DATA_RELOCATION_SRAM if XIP
+	help
+	  Enable the MCUX C40 internal flash driver used on NXP S32K3x
+	  (e.g. S32K344). Provides Zephyr flash driver using MCUX C40 HAL.
+
+# NXP MCUX C40 internal flash: optional protection pass at init
+config SOC_FLASH_MCUX_C40_APPLY_PROTECTION
```

Comment:
guard with `if SOC_FLASH_MCUX_C40` and drop `depends on SOC_FLASH_MCUX_C40` on this config. See how `config SOC_FLASH_MCUX` was done above, that's the correct approach.